File: ripbull-chat-sdk/src/main/java/com/ripbull/coresdk/user/UserModuleStub.java
package com.ripbull.coresdk.user;

import android.content.Context;
import androidx.annotation.NonNull;
import androidx.annotation.RestrictTo;
import com.ripbull.coresdk.R;
import com.ripbull.coresdk.core.ChatSDKException;
import com.ripbull.coresdk.core.type.AvailabilityStatus;
import com.ripbull.coresdk.core.type.ChatType;
import com.ripbull.coresdk.data.common.Result;
import com.ripbull.coresdk.eRTCSDK;
import com.ripbull.coresdk.user.mapper.UserMetaDataRecord;
import com.ripbull.coresdk.user.mapper.UserRecord;
import com.ripbull.coresdk.utils.Constants;
import io.reactivex.Flowable;
import io.reactivex.Observable;
import io.reactivex.Single;
import java.util.List;

/** @author meeth */
@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
public class UserModuleStub implements UserModule {

  private Context appContext;

  public static UserModule newInstance() {
    return new UserModuleStub(eRTCSDK.getAppContext());
  }

  private UserModuleStub(Context context) {
    this.appContext = context;
  }

  @Override
  public Flowable<List<UserRecord>> getChatUsers() {
    return Flowable.error(
        new ChatSDKException(new ChatSDKException.Error.InvalidModule(),
            appContext.getString(R.string.alert_message, Constants.Features.USER_MODULE)));
  }

  @Override
  public Single<List<UserRecord>> getMentionedUsers() {
    return Single.error(
            new ChatSDKException(new ChatSDKException.Error.InvalidModule(),
                    appContext.getString(R.string.alert_message, Constants.Features.USER_MODULE)));
  }

  @Override
  public Single<List<UserRecord>> getReactionedUsers(List<String> reactionUnicodes, String msgId , String threadId, ChatType chatType) {
    return Single.error(
            new ChatSDKException(new ChatSDKException.Error.InvalidModule(),
                    appContext.getString(R.string.alert_message, Constants.Features.USER_MODULE)));
  }

  @Override
  public Flowable<UserRecord> getUserById(@NonNull String id) {
    return Flowable.error(
        new ChatSDKException(new ChatSDKException.Error.InvalidModule(),
            appContext.getString(R.string.alert_message, Constants.Features.USER_MODULE)));
  }

  @Override
  public Single<Boolean> fetchMoreUsers() {
    return Single.error(
        new ChatSDKException(new ChatSDKException.Error.InvalidModule(),
            appContext.getString(R.string.alert_message, Constants.Features.USER_MODULE)));
  }

  @Override
  public Flowable<List<UserRecord>> getNewUsers(String addUpdateOrDelete) {
    return Flowable.error(
        new ChatSDKException(new ChatSDKException.Error.InvalidModule(),
            appContext.getString(R.string.alert_message, Constants.Features.USER_MODULE)));
  }

  @Override
  public Single<Result> logout() {
    return Single.error(
        new ChatSDKException(new ChatSDKException.Error.InvalidModule(),
            appContext.getString(R.string.alert_message, Constants.Features.LOGOUT)));
  }

  @Override
  public Single<Result> updateProfile(@NonNull String profileStatus, @NonNull String mediaPath,
      @NonNull String mediaType) {
    if (mediaPath == null || mediaPath.isEmpty()) {
      return Single.error(
          new ChatSDKException(new ChatSDKException.Error.InvalidModule(),
              appContext.getString(R.string.alert_message, Constants.Features.USER_PROFILE)));
    }
    return Single.error(
        new ChatSDKException(new ChatSDKException.Error.InvalidModule(),
            appContext.getString(R.string.alert_message, Constants.Features.USER_PROFILE_IMAGE)));
  }

  @Override
  public Single<UserRecord> getProfile() {
    return Single.error(
        new ChatSDKException(new ChatSDKException.Error.InvalidModule(),
            appContext.getString(R.string.alert_message, Constants.Features.USER_MODULE)));
  }

  @Override
  public Flowable<UserRecord> getLoggedInUser() {
    return Flowable.error(
        new ChatSDKException(new ChatSDKException.Error.InvalidModule(),
            appContext.getString(R.string.alert_message, Constants.Features.USER_MODULE)));
  }

  @Override
  public String getUserAvailabilityStatus() {
    return (appContext.getString(R.string.alert_message, Constants.Features.AVAILABILITY_STATUS));
  }

  @Override
  public Single setUserAvailability(AvailabilityStatus availabilityStatus) {
    return Single.error(
        new ChatSDKException(new ChatSDKException.Error.InvalidModule(),
            appContext.getString(R.string.alert_message, Constants.Features.AVAILABILITY_STATUS)));
  }

  @Override
  public Observable<UserRecord> subscribeToUserMetaData() {
    return Observable.error(
        new ChatSDKException(new ChatSDKException.Error.InvalidModule(),
            appContext.getString(R.string.alert_message, Constants.Features.USER_MODULE)));
  }

  @Override
  public Single<Result> removeProfilePic() {
    return Single.error(
        new ChatSDKException(new ChatSDKException.Error.InvalidModule(),
            appContext.getString(R.string.alert_message, Constants.Features.USER_PROFILE_IMAGE)));
  }

  @Override
  public Single<Result> deactivate() {
    return Single.error(
        new ChatSDKException(new ChatSDKException.Error.InvalidModule(),
            appContext.getString(R.string.alert_message, Constants.Features.LOGOUT)));
  }

  @Override
  public Observable<UserMetaDataRecord> metaDataOn(String appUserId) {
    return Observable.error(
        new ChatSDKException(new ChatSDKException.Error.InvalidModule(),
            appContext.getString(R.string.alert_message, Constants.Features.MUTE_NOTIFICATIONS)));
  }

  @Override
  public Single<Result> fetchLatestUserStatus() {
    return Single.error(
        new ChatSDKException(new ChatSDKException.Error.InvalidModule(),
            appContext.getString(R.string.alert_message, Constants.Features.AVAILABILITY_STATUS)));
  }

  @Override
  public Observable<Result> subscribeToLogout() {
    return null;
  }
}

File: InAppChat/src/main/java/io/inappchat/inappchat/user/UserModuleStub.kt
package io.inappchat.inappchat.user

import android.content.Context
import androidx.annotation.RestrictTo
import io.inappchat.inappchat.InAppChat.Companion.appContext
import io.inappchat.inappchat.R
import io.inappchat.inappchat.core.ChatSDKException
import io.inappchat.inappchat.core.ChatSDKException.Error.InvalidModule
import io.inappchat.inappchat.core.type.AvailabilityStatus
import io.inappchat.inappchat.core.type.ChatType
import io.inappchat.inappchat.data.common.Result
import io.inappchat.inappchat.user.mapper.UserMetaDataRecord
import io.inappchat.inappchat.user.mapper.UserRecord
import io.inappchat.inappchat.utils.Constants
import io.reactivex.rxjava3.core.Flowable
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.core.Single
import io.reactivex.rxjava3.subjects.Subject

/** @author meeth
 */
@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
class UserModuleStub private constructor(private val appContext: Context?) : UserModule {

    override val chatUsers: Flowable<List<UserRecord>>
        get() = Flowable.error(
            ChatSDKException(
                InvalidModule(),
                appContext!!.getString(R.string.alert_message, Constants.Features.USER_MODULE)
            )
        )
    override val mentionedUsers: Single<List<UserRecord>>
        get() = Single.error(
            ChatSDKException(
                InvalidModule(),
                appContext!!.getString(R.string.alert_message, Constants.Features.USER_MODULE)
            )
        )

    override fun getReactionedUsers(
        reactionUnicodes: List<String>,
        msgId: String,
        threadId: String,
        chatType: ChatType
    ): Single<List<UserRecord>> {
        return Single.error(
            ChatSDKException(
                InvalidModule(),
                appContext!!.getString(R.string.alert_message, Constants.Features.USER_MODULE)
            )
        )
    }

    override fun getUserById(id: String): Flowable<UserRecord> {
        return Flowable.error(
            ChatSDKException(
                InvalidModule(),
                appContext!!.getString(R.string.alert_message, Constants.Features.USER_MODULE)
            )
        )
    }

    override fun fetchMoreUsers(): Single<Boolean> {
        return Single.error(
            ChatSDKException(
                InvalidModule(),
                appContext!!.getString(R.string.alert_message, Constants.Features.USER_MODULE)
            )
        )
    }

    override fun getNewUsers(addUpdateOrDelete: String): Flowable<List<UserRecord>> {
        return Flowable.error(
            ChatSDKException(
                InvalidModule(),
                appContext!!.getString(R.string.alert_message, Constants.Features.USER_MODULE)
            )
        )
    }

    override fun logout(): Single<Result> {
        return Single.error(
            ChatSDKException(
                InvalidModule(),
                appContext!!.getString(R.string.alert_message, Constants.Features.LOGOUT)
            )
        )
    }

    override fun updateProfile(
        profileStatus: String, mediaPath: String,
        mediaType: String
    ): Single<Result> {
        return if (mediaPath == null || mediaPath.isEmpty()) {
            Single.error(
                ChatSDKException(
                    InvalidModule(),
                    appContext!!.getString(
                        R.string.alert_message,
                        Constants.Features.USER_PROFILE
                    )
                )
            )
        } else Single.error(
            ChatSDKException(
                InvalidModule(),
                appContext!!.getString(
                    R.string.alert_message,
                    Constants.Features.USER_PROFILE_IMAGE
                )
            )
        )
    }

    override val profile: Single<UserRecord>
        get() = TODO("Not yet implemented")
    override val loggedInUser: Flowable<UserRecord>
        get() = TODO("Not yet implemented")
    override val userAvailabilityStatus: String
        get() = TODO("Not yet implemented")

    override fun setUserAvailability(availabilityStatus: AvailabilityStatus): Single<*> {
        return Single.error<Any>(
            ChatSDKException(
                InvalidModule(),
                appContext!!.getString(
                    R.string.alert_message,
                    Constants.Features.AVAILABILITY_STATUS
                )
            )
        )
    }

    override fun subscribeToUserMetaData(): Observable<UserRecord> {
        return Observable.error(
            ChatSDKException(
                InvalidModule(),
                appContext!!.getString(R.string.alert_message, Constants.Features.USER_MODULE)
            )
        )
    }

    override fun removeProfilePic(): Single<Result> {
        return Single.error(
            ChatSDKException(
                InvalidModule(),
                appContext!!.getString(
                    R.string.alert_message,
                    Constants.Features.USER_PROFILE_IMAGE
                )
            )
        )
    }

    override fun deactivate(): Single<Result> {
        return Single.error(
            ChatSDKException(
                InvalidModule(),
                appContext!!.getString(R.string.alert_message, Constants.Features.LOGOUT)
            )
        )
    }

    override fun metaDataOn(appUserId: String): Observable<UserMetaDataRecord> {
        return Observable.error(
            ChatSDKException(
                InvalidModule(),
                appContext!!.getString(
                    R.string.alert_message,
                    Constants.Features.MUTE_NOTIFICATIONS
                )
            )
        )
    }

    override fun fetchLatestUserStatus(): Single<Result> {
        return Single.error(
            ChatSDKException(
                InvalidModule(),
                appContext!!.getString(
                    R.string.alert_message,
                    Constants.Features.AVAILABILITY_STATUS
                )
            )
        )
    }

    private val logoutSubject = Subject.empty<Result>()
    override fun subscribeToLogout(): Observable<Result> {
        return logoutSubject
    }

    companion object {
        @JvmStatic
        fun newInstance(): UserModule {
            return UserModuleStub(appContext)
        }
    }
}